### ISSUE TITLE
Prefer QUIC for WAN connections

### ISSUE THREAD

**Issue by bt90** [syncthing/syncthing#8626]
This is a bit of a long shot and depends on the performance of future quic-go releases but QUIC already tends to perform better on high latency connections:

https://forum.syncthing.net/t/prefer-quic-option/19136

https://www.reddit.com/r/Syncthing/comments/ye0429/syncthing_quicserver_quickclient_disconnecting/

Seems like quic-go [v0.30.0](https://github.com/lucas-clemente/quic-go/releases/tag/v0.30.0) already comes with a bunch of performance improvements and [batch writing](https://github.com/lucas-clemente/quic-go/pull/3524) is scheduled for v0.31.0.

Once quic-go is able to saturate a gigabit link, we should change the priority for WAN connections.

As for LAN i don't see any benefits because packetloss and latency is not an issue here.

**Comment by calmh** [comment]
We might want to consider some sort of circuit breaker if implementing this, to prevent an eternal connection-upgrade fight between two differently configured devices.

**Comment by bt90** [comment]
At least to avoid issues with nodes running an older Syncthing version.

I wonder if we can already drop the QUIC->TCP upgrade logic.

**Comment by AudriusButkevicius** [comment]
I also think the "gigabits" claim needs to be verified. 
We have a lot of glue around quic, which might be the bottleneck achieving the said gigabytes.
We don't need that glue for TCP.

Also, if anything, the batch writes might break the existing glue rather than make things better.

**Comment by bt90** [comment]
There's also been some discussion on caddys issue tracker around the performance of quic-go: https://github.com/caddyserver/caddy/issues/4876

As you're mentioning the glue code support:  https://github.com/AudriusButkevicius/pfilter/issues/7

**Comment by The0013** [comment]
I am sorry that I don't understand a thing from what you guys are talking about! but it looks like there is no solution for my problem for now? Can anyone at least know what is the magical number 1.6 mib/s that I am stuck at?
![image](https://user-images.githubusercontent.com/117287222/199504591-ac310353-75dc-43a8-8f5c-3d2b4e82eca6.png)
the line looks flat, as if there is something caping the tcp connection! or is there a way to block the tcp connection I tried adding the tcp port to the firewall, but that did not prevent it from connecting don't know why

**Comment by calmh** [comment]
@The0013 your issue is something else, please seek help to debug it on the forum.

**Comment by bt90** [comment]
Looks like quic-go is fast enough for our purposes and is already aiming to be able to saturate 10Gbps links: https://github.com/lucas-clemente/quic-go/issues/3670

**Comment by imsodin** [comment]
Yeah, the bare quic-go implementation is performant enough based on those reports. There's still the glue around it that Audrius mentioned plus real networks: I'd be surprised if UDP traffic works as well (more so on reliability than throughput) in the wild. Not because it couldn't, but because of old hard/software or policies/firewalls getting in the way. I'd be vary of changing the default, and maybe make it an opt-in config first. Plus the mentioned problem of flip-flop-management.

**Comment by bt90** [comment]
I think the safest option would be to treat QUIC and TCP as equal, but choose QUIC first. This would avoid the flip-flop problem with older nodes. If only new nodes are involved and a QUIC connection could be established, it stays that way. An older device could force TCP, but again this would not cause any problems because the newer nodes would see no reason to switch it to QUIC.

So instead of:

Relay < QUIC WAN < TCP WAN < TCP LAN

We would treat it this way:

Relay < QUIC WAN == TCP WAN < TCP LAN

I'm not that concerened about UDP problems to be honest. The adoption of HTTP/3 forces things to work at least on the provider level. If it's blocked in corporate networks or public wifis we can still fallback to TCP.

**Comment by bt90** [comment]
Blocked by #7403

**Comment by daiaji** [comment]
QoS on older gateways is actually less of a concern, as HTTP3 advances they have to accommodate large amounts of UDP data.

**Comment by pataquets** [comment]
> This is a bit of a long shot and depends on the performance of future quic-go releases but QUIC already tends to perform better on high latency connections:
> 
> https://forum.syncthing.net/t/prefer-quic-option/19136
> 
> https://www.reddit.com/r/Syncthing/comments/ye0429/syncthing_quicserver_quickclient_disconnecting/
> 
> Seems like quic-go [v0.30.0](https://github.com/lucas-clemente/quic-go/releases/tag/v0.30.0) already comes with a bunch of performance improvements and [batch writing](https://github.com/lucas-clemente/quic-go/pull/3524) is scheduled for v0.31.0.
> 
> Once quic-go is able to saturate a gigabit link, we should change the priority for WAN connections.
> 

+1

> As for LAN i don't see any benefits because packetloss and latency is not an issue here.

Just wondering... QUIC (aka UDP) on LAN would be easy on network equipment which might come handy if there's lots of traffic. Also, being UDP easily dropped in need, it would coexist better with TCP streaming and low latency applications.
Just my 2c.

**Comment by calmh** [comment]
On further consideration I think this is resolved as you can now tell Syncthing to prefer QUIC, though it's not the default.

**Comment by pataquets** [comment]
@calmh How do you tell ST to prefer QUIC? Is there any way to make it default (eg. a command line switch)?
I can file another issue, if required.

**Comment by calmh** [comment]
@pataquets by setting the connection priority for QUIC to be lower than TCP. Advanced options. I should write a doc article one of these days.

**Comment by The0013** [comment]
THANK YOU VERY MUCH

**Comment by nikow** [comment]
![image](https://github.com/syncthing/syncthing/assets/4789807/71f8726c-b8a3-4986-a2d3-5e48951c3958)
Looks like we just need to tweak those numbers. Thank you very much!
